pkg/provider/ip/provider.go

```diff
@@ -120,11 +154,52 @@ func (p *ipv4Provider) DeInitResource(instance ec2.EC2Instance) error {
 
 // UpdateResourceCapacity updates the resource capacity based on the type of instance
 func (p *ipv4Provider) UpdateResourceCapacity(instance ec2.EC2Instance) error {
+	resourceProviderAndPool, isPresent := p.getInstanceProviderAndPool(instance.Name())
+	if !isPresent {
+		p.log.Error(nil, "cannot find the instance provider and pool form the cache", "node-name", instance.Name())
+		return nil
+	}
+
+	resourceProviderAndPool.lock.Lock()
+	defer resourceProviderAndPool.lock.Unlock()
+
+	// Check if PD is enabled
+	isCurrPDEnabled := p.conditions.IsWindowsPrefixDelegationEnabled()
+
+	// Previous state and current state are either both enabled or disabled, effectively no change
+	if resourceProviderAndPool.isPrevPDEnabled == isCurrPDEnabled {
+		return nil
+	}
+
+	// If prefix delegation is enabled, then set the secondary IP pool state to draining and
+	// do not update the capacity as that would be done by prefix provider
+	if !resourceProviderAndPool.isPrevPDEnabled && isCurrPDEnabled {
+		resourceProviderAndPool.isPrevPDEnabled = true
+
+		job := resourceProviderAndPool.resourcePool.SetToDraining()
+		p.log.Info("IPv4 prefix provider should be active")
+		p.SubmitAsyncJob(job)
+		return nil
+	}
+
+	resourceProviderAndPool.isPrevPDEnabled = false
+
+	// Set the secondary IP provider pool state to active
+	resourceConfig := config.LoadResourceConfig()
+	ipv4ResourceConfig, ok := resourceConfig[config.ResourceNameIPAddress]
+	if !ok {
+		p.log.Error(fmt.Errorf("failed to find resource configuration"), "update resource capacity failed",
+			"resource name", config.ResourceNameIPAddress)
+		return nil
+	}
+	job := resourceProviderAndPool.resourcePool.SetToActive(ipv4ResourceConfig.WarmPoolConfig)
+	p.SubmitAsyncJob(job)
+
 	instanceType := instance.Type()
 	instanceName := instance.Name()
 	os := instance.Os()
 
-	capacity := getCapacity(instanceType, os)
+	capacity := resourceProviderAndPool.capacity
```

Comment:
@haouc `resourceProviderAndPool.capacity` is being read here so the lock is required as well.

Reply:
Is that possible to have RLocks in the calling functions instead of the caller? For example, in functions like GetCapacity()?

Reply:
I see what you are saying. The reason I added here is the lock was already acquired at the beginning of this function and deferred releasing since it was needed for updating `isPDEnabled` so I thought not to acquire it again. This is all part of `UpdateResourceCapacity` function, but github added a few automatic checks that broke down the block into fragments which made it more confusing for reviewing. Sorry about that!

Reply:
We can have the lock like this for now. When we have the entire package ready (the bigger PR), we can revisit the locks. We definitely need to avoid recall lock!